File: path-sum-iii/path-sum-iii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    # def pathSum(self, root: Optional[TreeNode], targetSum: int) -> int:
    def pathSum(self, root, target):
                          # define global result and path
        self.result = 0
        distance = {0:1}          
        self.dfs(root, target, 0, distance)       # recursive to get result                      
        return self.result   # return result
    
    def dfs(self, root, target, currPathSum, distance): # 中 左 右 顺序进行遍历
                           
        if root is None:   # exit condition
            return  
                     # calculate currPathSum and required oldPathSum
        currPathSum += root.val
        oldPathSum = currPathSum - target
                     # update result and distance
        self.result += distance.get(oldPathSum, 0)
        distance[currPathSum] = distance.get(currPathSum, 0) + 1     #更新频率
        
                    # dfs breakdown
        self.dfs(root.left, target, currPathSum, distance)
        self.dfs(root.right, target, currPathSum, distance)
                     # when move to a different branch, 
                     #the currPathSum is no longer available, hence remove one. 
        distance[currPathSum] -= 1     

File: path-sum-iii/path-sum-iii.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int pathSum(TreeNode root, int sum) {
        if (root == null) {
            return 0;
        }

      //1.由于它可以不从根节点出发，因此我们可以这么写
        return paths(root, sum) 
                + pathSum(root.left, sum) 
                + pathSum(root.right, sum);
    }

    private int paths(TreeNode root, int sum) {

        if (root == null) {
            return 0;
        }

        int res = 0;
        if (root.val == sum) {
            res += 1;            
        }
        
      
      //2.通过这里保证连续性
        res += paths(root.left, sum - root.val);
        res += paths(root.right, sum - root.val);
        
        return res;
    }
}